# Исследование надежности заемщиков

## 1. Откроем таблицу и изучим общую информацию о данных

**1.1 Импортируем библиотеку pandas. Считаем данные из csv - файла в датафрейм и сохраним в переменную data.**

In [1]:
import pandas as pd

In [3]:
data = pd.read_csv('C:/Users/RYaku/OneDrive/Рабочий стол/МОИ ПРОЕКТЫ/Яндекс практикум/Второй проект/data.csv')

**1.2 Выведим первые 20 строчек датафрейма data на экран**

In [5]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**1.3 Выведиv основную информацию о датафрейме с помощью метода info().**

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## 2. Предобработка данных

### 2.1 Удаление пропусков

**2.1.1 Выведим количество пропущенных значений для каждого столбца. Используем комбинацию двух методов.**

In [8]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**2.1.2 total_income — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца income_type**

In [9]:
data['total_income'].fillna(data.groupby('income_type')['total_income'].transform('median'), inplace=True)
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
        data.loc[data['income_type'] == t, 'total_income'].median()

### 2.2 Обработка аномальных значений

**2.2.1 В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. Таким артефактом будет отрицательное количество дней трудового стажа в столбце days_employed. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода abs().**

In [13]:
data['days_employed'] = data['days_employed'].abs()

**2.2.2 Для каждого типа занятости выведим медианное значение трудового стажа days_employed в днях.**

In [15]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

**2.2.3 Выведим перечень уникальных значений столбца children**

In [16]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**2.2.4 В столбце children есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма data.**

In [17]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**2.2.5 Снова выведим перечень уникальных значений столбца children**

In [18]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### 2.3 Удаление пропусков (продолжение)

**2.3.1 Заполним пропуски в столбце days_employed медианными значениями по каждому типу занятости income_type.**

In [19]:
for income_type in data['income_type'].unique():
    median_days_employed = data.loc[data['income_type'] == income_type, 'days_employed'].median()
    data.loc[(data['income_type'] == income_type) & (data['days_employed'].isna()), 'days_employed'] = median_days_employed

**2.3.2 Убедимся, что все пропуски заполнены.**

In [21]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

In [22]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21402 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21402 non-null  int64  
 1   days_employed     21402 non-null  float64
 2   dob_years         21402 non-null  int64  
 3   education         21402 non-null  object 
 4   education_id      21402 non-null  int64  
 5   family_status     21402 non-null  object 
 6   family_status_id  21402 non-null  int64  
 7   gender            21402 non-null  object 
 8   income_type       21402 non-null  object 
 9   debt              21402 non-null  int64  
 10  total_income      21402 non-null  float64
 11  purpose           21402 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


### 2.4 Изменение типов данных

**2.4.1 Заменим вещественный тип данных в столбце total_income на целочисленный с помощью метода astype().**

In [23]:
data['total_income'] = data['total_income'].astype(int)

### 2.5 Обработка дубликатов

**2.5.1 Обработаем неявные дубликаты в столбце education. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру.**

In [24]:
data['education'] = data['education'].str.lower()

**2.5.2 Выведим на экран количество строк-дубликатов в данных.**

In [25]:
data.duplicated().sum()

71

In [27]:
data = data.drop_duplicates()

**2.5.3 Снова выведим на экран количество строк-дубликатов в данных.**

In [28]:
data.duplicated().sum()

0

### 2.6 Категоризация данных

**2.6.1  На основании диапазонов, указанных ниже, создадим в датафрейме data столбец total_income_category с категориями:**

- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.
 
**Например, кредитополучателю с доходом 25000 нужно назначить категорию 'E', а клиенту, получающему 235000, — 'B'. Используйте собственную функцию с именем categorize_income() и метод apply().**

In [39]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [31]:

data['total_income_category'] = data['total_income'].apply(categorize_income)

**2.6.2 Проверим созданный столбец**

In [34]:
print(data.columns)

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose', 'total_income_category'],
      dtype='object')


**2.6.3 Выведим на экран перечень уникальных целей взятия кредита из столбца purpose.**

In [32]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**2.6.4 Создадим функцию, которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образовани'`.я'.

def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [36]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### 2.7 Шаг 3. Исследуем вопросы

#### 2.7.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [77]:
def calculate_mean_percentage(x):
    return round(x.mean() * 100, 2)

result = data.groupby('children')['debt'].agg(['count', calculate_mean_percentage])
result.columns = ['количество_кредитов', 'доля_просрочек']

result

количество_кредитов  доля_просрочек
children                                     
0                       14091            7.54
1                        4808            9.23
2                        2052            9.45
3                         330            8.18
4                          41            9.76
5                           9            0.00

**Исходя из представленных данных, можно сделать следующие выводы:**

- 1. Группа заемщиков без детей (0 детей) имеет наименьшую долю невозвратов кредита - 7,54%. Это может указывать на то, что отсутствие детей позволяет более стабильно управлять финансами и своевременно возвращать кредиты.

- 2. С увеличением количества детей у заемщиков, наблюдается некоторое увеличение доли невозвратов кредита. Так, для групп с 1, 2 и 4 детьми доля невозвратов составляет 9,23%, 9,45% и 9,76% соответственно.

- 3. Однако стоит отметить, что группа с 3 детьми имеет более низкую долю невозвратов (8,18%) по сравнению с группами с 1, 2 и 4 детьми. Это может указывать на то, что влияние количества детей на возврат кредита не является строго линейным.

- 4. Группа с 5 детьми имеет 0% невозвратов, но размер выборки для этой группы очень мал (всего 9 заемщиков), поэтому делать однозначные выводы по ней не представляется возможным.

**В целом, данные указывают на наличие некоторой зависимости между количеством детей и своевременным возвратом кредита, однако эта зависимость не является прямолинейной. Необходимо дальнейшее изучение данного вопроса с привлечением более обширной выборки.** 

#### 2.7.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [64]:
result = data.groupby(['family_status','family_status_id'])['debt'].agg(
    количество_кредитов='count', 
    доля_просрочек ='mean'
).sort_values('family_status_id')

result['доля_просрочек'] = round(result['доля_просрочек'] * 100, 2)
result

количество_кредитов  доля_просрочек
family_status         family_status_id                                     
женат / замужем       0                               12261            7.56
гражданский брак      1                                4134            9.31
вдовец / вдова        2                                 951            6.62
в разводе             3                                1189            7.06
Не женат / не замужем 4                                2796            9.76

**Исходя из представленных данных, можно сделать вывод, что существует некоторая зависимость между семейным положением и возвратом кредита в срок. Процент невозврата кредита (доля просрочек) варьируется в зависимости от семейного статуса:**

- Люди, находящиеся в браке ("женат / замужем"), имеют более низкий процент просрочек (7.56%).
- Те, кто находится в гражданском браке или не женат / не замужем, демонстрируют более высокий уровень просрочек (9.31% и 9.76% соответственно).
- Вдовцы и вдовы, а также люди в разводе, показывают относительно низкую долю просрочек (6.62% и 7.06% соответственно).

**Эти данные могут указывать на то, что люди, которые не состоят в официальном браке, могут иметь более высокий риск просрочки платежей по кредитам. Однако для окончательных выводов необходимо провести более глубокий статистический анализ, так как могут быть и другие факторы, влияющие на возврат кредита в срок.**

#### 2.7.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [73]:
result = data.groupby(['total_income_category'])['debt'].agg(
    количество_кредитов='count', 
    доля_просрочек ='mean').sort_values('доля_просрочек')
result['доля_просрочек'] = round(result['доля_просрочек'] * 100, 2)
result

количество_кредитов  доля_просрочек
total_income_category                                     
D                                      349            6.02
B                                     5014            7.06
A                                       25            8.00
C                                    15921            8.50
E                                       22            9.09

**Исходя из представленных данных, можно сделать следующие выводы, что существует некоторая зависимость между уровнем дохода и возвратом кредита в срок. Процент просрочек увеличивается с повышением категории дохода. Похоже, что люди с самым низким и самым высоким уровнем дохода более склонны к просрочкам по кредитам по сравнению с теми, кто находится в средней категории дохода**

#### 2.7.4 Как разные цели кредита влияют на его возврат в срок?

In [76]:
result = data.groupby(['purpose_category'])['debt'].agg(
    количество_кредитов='count', 
    доля_просрочек ='mean').sort_values('доля_просрочек')
result['доля_просрочек'] = round(result['доля_просрочек'] * 100, 2)
result

количество_кредитов  доля_просрочек
purpose_category                                             
операции с недвижимостью                10751            7.26
проведение свадьбы                       2313            7.91
получение образования                    3988            9.25
операции с автомобилем                   4279            9.35

**На основе представленных данных можно сделать краткий вывод, что цели кредита влияют на вероятность его возврата в срок. Кредиты на операции с недвижимостью имеют наименьший процент просрочек, что может свидетельствовать о более высокой ответственности заемщиков в этой категории. В то время как кредиты на получение образования и операции с автомобилем демонстрируют более высокую долю просрочек, что указывает на потенциально больший риск невозврата для кредиторов.**

#### 2.7.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* *Пропуски в исходных данных могут возникнуть по различным причинам, которые не обязательно связаны с намеренным уклонением от предоставления информации. Возможные объяснения включают технические ошибки при сборе или передаче данных, человеческий фактор, такой как невнимательность или забывчивость при заполнении форм, а также ограничения системы регистрации данных, которая может не требовать обязательного заполнения всех полей. Для получения кредита, особенно на недвижимость, обычно требуется предоставление полного пакета информации, поэтому если пропуски присутствуют, они, скорее всего, являются случайными, а не результатом умышленного уклонения.*

#### 2.7.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ: Использование медианы для заполнения пропусков в количественных переменных является предпочтительным, поскольку медиана менее подвержена влиянию экстремальных значений или выбросов в данных. В отличие от среднего, медианное значение не искажается из-за нескольких очень высоких или очень низких значений, что делает его более надежным показателем центральной тенденции для асимметрично распределенных данных. Это означает, что медиана может быть использована в качестве универсального метода для заполнения пропусков, тем самым минимизируя искажение общей структуры данных. Однако важно учитывать контекст данных и не заполнять все пропуски одной и той же медианой для всего набора данных, так как это может внести искажения. Лучше использовать медианы, рассчитанные для более мелких и однородных групп данных, чтобы сохранить их структурную целостность.*

### Шаг 4: общий вывод.

*В рамках исследования кредитоспособности заемщиков был проведен анализ предоставленных данных. Процесс включал очистку данных от пропусков и дубликатов, а также классификацию клиентов по ключевым характеристикам. На основе анализа можно сделать следующие обобщения:*

*Влияние наличия детей на возврат кредита:*

*Заемщики без детей демонстрируют лучшую историю по возврату кредитов, что может быть связано с отсутствием дополнительных расходов на содержание детей. В то же время, информация о заемщиках с пятью детьми ограничена небольшим числом наблюдений, что затрудняет делать обоснованные выводы для этой категории.*

*Связь семейного статуса с погашением кредитов:*

*Наблюдается зависимость между семейным статусом заемщика и его кредитной дисциплиной. Люди, не состоящие или никогда не состоявшие в официальном браке, чаще допускают просрочки по кредитам, в то время как вдовцы и вдовы, возможно благодаря большей жизненной ответственности и поддержке со стороны взрослых детей, имеют наименьший процент просрочек.*

*Зависимость между уровнем дохода и своевременностью возврата кредита:*

*Явной связи между доходом заемщика и вероятностью просрочек не выявлено. Размеры групп по уровню дохода слишком различаются, чтобы делать однозначные выводы.*

*Влияние целей кредита на его возврат в срок:*

*Цели кредита оказывают влияние на дисциплину его возврата. Кредиты, взятые на приобретение недвижимости, обладают наименьшим процентом просрочки, возможно, из-за их воспринимаемой заемщиками важности и необходимости. В то же время кредиты на покупку автомобиля чаще всего не погашаются в срок, что может быть связано с более импульсивной природой таких покупок или с непредвиденными расходами на эксплуатацию авто.*